### ISSUE TITLE
Support for disabling type checks of type alternatives

### ISSUE THREAD

**Issue by Flamefire** [microsoft/pyright#9945]
**Is your feature request related to a problem? Please describe.**

I'm implementing a Web-API call that returns JSON data. For that I have a generic function that does the request, handles the error and returns the data on success. The type of the data depends on the API called. Say `request('get/id')` returns an `int` and `request('get/name')` returns a string. There might even be some that do not return anything, i.e. `None`.

I now implement this as e.g.:
```
def request(endpoint: str) -> str | int | None:
    # Do it and return

def get_id() -> int:
    return request('get/id')

def get_name() -> str:
    return request('get/name')
```

Similar the result might be a JSON structure specific to each endpoint where keys are strings and values either int, string or list of strings.
Then I'd do something like: `token = TokenData(created=data['created'], num_uses=data['num_uses'], users=data['users'])`   
From the endpoint description I know the types are `str, int, list[str]` respectively and want to work with a dataclass now instead of the JSON dict

The issue with the type checker is now that each time a "type narrowing", e.g. `"str | int" -> "int"` or `str | list[str] -> list[str]` occurs that is logically correct, the type checker (correctly) warns that not all source types are compatible with the target type. This happens for return values, function arguments and variable assignments.

Putting `isinstance` assertions doesn't work well for any of those cases and using `cast` everywhere is cumbersome and overly verbose

**Describe the solution you’d like**

Add an option to the checker that disables the `reportReturnType` diagnostics if any alternative is assignable. Possibly as a macro for function and file scope for e.g. "low-level library code"

Basically I want the inverse of being able to return an `int` for a return type of `int | str`.

This still allows to catch instances that are definitely wrong, e.g. `list -> int` and `str | int -> list` at the expense of missing some where the implicit assumptions don't hold.    
But this makes the type checks much easier to use as e.g. this lenient check can be enabled without much work and then refined using the stricter checks later if necessary. Especially with the function/file level macros it could be limited to the plumbing while business logic uses the strict checks.

**Comment by erictraut** [comment]
It's unlikely we'd add such a feature. The assignability rules are well defined in the typing spec. What you're asking for is to support a modified assignability rule in a particular case. There is no precedent for deviating from standard assignability rules via a configuration option, and I don't think such a rule would be a good idea. Code that uses such a rule would not be very robust.

My recommendation is to write a set of validation functions that perform a runtime validation (perhaps using an `isinstance` check or additional checks if necessary).

```python
def request(endpoint: str) -> str | int | None:
    # Do it and return
    pass

def verify_int(value: str | int | None) -> int:
    assert isinstance(value, int)
    return value

def verify_str(value: str | int | None) -> str:
    assert isinstance(value, str)
    return value

def get_id() -> int:
    return verify_int(request("get/id"))

def get_name() -> str:
    return verify_str(request("get/name"))
```
